### PR TITLE
Refactor rename derivation type `isPure`

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

To quote the method doc:

Non-impure derivations can still behave impurely, to the degree permitted
by the sandbox. Hence why this method isn't `isPure`: impure derivations
are not the negation of pure derivations. Purity can not be ascertained
except by rather heavy tools.

Changes:

- Partially correct this mistake in reasoning. _Even if it's not that much less similar to "no purity", it at least has more similarity to the feature name._
- Add _impure derivation_ to the glossary.

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by edolstra** [comment]
I don't think "impure" is better than "pure". After all, it still implies that a "not impure" derivation is "pure". The reality is that Nix's notion of "purity" in the context of derivations is not the same as mathematical purity, and we should just document that (or come up with another term).

Also, double negations like `!isImpure()` are less readable.

**Comment by roberth** [comment]
> t still implies that a "not impure" derivation is "pure"

Only if you believe in the law of excluded middle. Purity is basically undecidable, so we don't really have that luxury.



> double negations like `!isImpure()` are less readable.

I'd prefer something less readable of something semantically inconsistent. `isPure()` would be wrong because it doesn't assert purity; only indicates that we're going to assume it.

> Nix's notion of "purity" in the context of derivations is not the same as mathematical purity

This seems really bad. How should users then describe derivations they write that do behave as pure functions?
Is there a fundamental reason why it has to be different?

We already have a somewhat different interpretation of reproducibility when we say that Nix guarantees it. Having to explain that we define things differently undermines our credibility, to a degree.

**Comment by Ericson2314** [comment]
Maybe a name like "expiring" would be better. In principle, whether we want to sandbox, and whether the "(drv, output) -> CA" realization is permanent or ephemeral are orthogonal.

File: doc/manual/src/glossary.md [hunk 1]
```diff
@@ -271,6 +276,21 @@
...
+
+- [package attribute set]{#package-attribute-set}
+
+  An [attribute set] containing the attribute `type = "derivation";` (derivation for historical reasons), as well as other attributes, such as
+  - attributes that refer to the files of a [package], typically in the form of [derivation outputs](#output),
+  - attributes that declare something about how the package is supposed to be installed or used,
+  - other metadata or arbitrary attributes.
+
+  [package attribute set]: #package-attribute-set
+
```
Comment (Ericson2314) [inline doc/manual/src/glossary.md:293]: Should go away with rebase

**Review by Ericson2314 — APPROVED** [review]

**Comment by thufschmitt** [comment]
Discussed during the Nix maintainers meeting on 2024-01-22. Idea approved.

<details>
- @tomberek & @ericson2314: if we consistently call thing thing "impure", the identifiers makes sense without even knowing "impure means not pure" in English. Seems good.
- @edolstra seems like churn for little/no gain, but I won't block it.

Approved (though needs rebase).

**Comment by roberth** [comment]
I have to admit that this PR was at the edge of what's acceptable for what is superficially a nitpick.

Nonetheless it is important to keep our terminology pure (pun intended), so I'll finish this.

**Pull request merged**